### ISSUE TITLE
When a `Literal` is used for a generic and wrapped into a `Callable`, it is treated as a `str`

### ISSUE THREAD

**Issue by wch** [microsoft/pyright#4996]
I have a function that serves as a wrapper: it takes either a value or a function that returns a value, and returns a function that returns the value. It has a signature like this:

```py
def wrap_function(x: T | Callable[[], T]) -> Callable[[], T]:
    ...
```

When I pass in something that is `str | Callable[[], str]`, pyright correctly infers that this returns a `Callable[[], str]`.

However, when I pass in a `Foo | Callable[[], Foo]`, where `Foo` is a `Literal`, then pyright thinks that it returns a `Callable[[], str]` -- but I think it should infer that the return value is `Callable[[], Foo]`.

Example:

```py
from typing import Callable, Literal, TypeVar

Foo = Literal["a", "b", "c"]
T = TypeVar("T")


def wrap_function(x: T | Callable[[], T]) -> Callable[[], T]:
    if not callable(x):
        return lambda: x

    return x


def test(
    x: str | Callable[[], str],
    y: Foo | Callable[[], Foo],
) -> None:
    x = wrap_function(x)  # OK
    y = wrap_function(y)  # Flagged
```

```
$ pyright test.py 
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.304
/Users/winston/test/test.py
  /Users/winston/test/test.py:19:9 - error: Expression of type "() -> str" cannot be assigned to declared type "Foo | (() -> Foo)"
    Type "() -> str" cannot be assigned to type "Foo | (() -> Foo)"
      Type "() -> str" cannot be assigned to type "() -> Foo"
        Function return type "str" is incompatible with type "Foo"
          Type "str" cannot be assigned to type "Foo"
            "str" cannot be assigned to type "Literal['a']"
            "str" cannot be assigned to type "Literal['b']"
            "str" cannot be assigned to type "Literal['c']"
      "function" cannot be assigned to type "Literal['a']"
    ... (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.463sec
```

**Comment by erictraut** [comment]
Literals are not generally retained by pyright's constraint solver unless they are required to satisfy the constraints. One common way to provide those constraints is through the use of bidirectional type inference, where the "expected type" requires a literal. Your example uses bidirectional type inference, but the "expected type" is a union of two types (`Foo` and `Callable[[], Foo]`). In some cases, pyright is able to filter the list of subtypes in a expected type union and use only the filtered type for bidirectional type inference, but there are cases that it cannot handle. This is one of them.

Note that if you change the code as follows, it works without error:

```python
def test(
    x: str | Callable[[], str],
    y: Foo | Callable[[], Foo],
) -> None:
    x = wrap_function(x)  # OK
    z: Callable[[], Foo] = wrap_function(y)  # OK
```

I'll see if I can improve the logic for filtering union subtypes in an "expected type" so it can handle this case.

**Comment by wch** [comment]
I did notice that I could add the annotation:

```py
z: Callable[[], Foo] = wrap_function(y)
```

As I'm sure you know, the drawback is that you have to use a new variable, `z`, since if you use `y` instead, pyright complains about it being redeclared.

*****

Another observation: I think a similar (or maybe the same?) problem happens even without a union. For example:

```py
from typing import Callable, Literal, TypeVar

Foo = Literal["a", "b", "c"]
T = TypeVar("T")

def wrap(x: T) -> Callable[[], T]:
    return lambda: x

def f_str(x: Callable[[], str]) -> str:
    return x()


def f_Foo(x: Callable[[], Foo]) -> Foo:
    return x()


def test(
    x: str,
    y: Foo,
) -> None:
    x1 = wrap(x)
    y1 = wrap(y)

    f_str(x1)  # OK
    f_Foo(y1)  # Flagged
```

```
$ pyright test.py
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.304
/Users/winston/test.py
  /Users/winston/test.py:28:11 - error: Argument of type "() -> str" cannot be assigned to parameter "x" of type "() -> Foo" in function "f_Foo"
    Type "() -> str" cannot be assigned to type "() -> Foo"
      Function return type "str" is incompatible with type "Foo"
        Type "str" cannot be assigned to type "Foo"
          "str" cannot be assigned to type "Literal['a']"
          "str" cannot be assigned to type "Literal['b']"
          "str" cannot be assigned to type "Literal['c']" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.453sec
```

**Comment by erictraut** [comment]
Yes, this is expected behavior. When evaluating the statement `y1 = wrap(y)`, the value `y1` has no declared value, so it imposes no constraints on the type of the expression `wrap(y)`, and the constraint solver does not retain the literal value of `y` — just as you wouldn't expect it to retain the literal value if you were to call `wrap('a')`.

**Comment by wch** [comment]
But shouldn't it know that `y1` is a `Callable[[], Foo]` because in that case , the `T` is `Foo`? Pylance thinks that it's a `Callable[[], str]`.

<img width="249" alt="image" src="https://user-images.githubusercontent.com/86978/233858222-c7610e16-5fa0-4018-a543-2a0161c75574.png">

FWIW, when I run `mypy test.py --strict` (with mypy 1.2.0) it says:

```
$ mypy test.py --strict 
Success: no issues found in 1 source file
```

(Not that I necessarily take mypy as correct, of course.)

**Comment by erictraut** [comment]
Yes, mypy is inconsistent in how it handles literals in its constraint solver. We've chosen to be consistent here. See [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constraint-solver-literals) for details.

I've added additional logic to the call expression evaluator so it can apply bidirectional type inference in cases where the return type of the call is a callable and the expected type is a union that contains a combination of callable and non-callable types. It now filters out the non-callables (such as `Foo` in your example above) which allows it to apply bidirectional type inference. This will be addressed in the next release.

**Comment by wch** [comment]
Thanks for addressing the original issue, and thanks for the pointer to the documentation about literals.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.305, which I just published. It will also be included in a future release of pylance.
